StreamReader.cpp

```diff
@@ -90,5 +79,6 @@ void MemoryStreamReader::SeekRelative(int64_t offset)
 		throw std::runtime_error("Change in offset places read position outside bounds of buffer.");
 	}
 
-	this->position += offset;
+	// offset is checked against size of streamSize, which cannot exceed SIZE_MAX (max size of size_t)
+	this->position += static_cast<size_t>(offset); 
```

Comment:
I believe there is some potential for bounds issues here. It's not `offset`, but rather the sum `position + offset` which is checked against `streamSize`. There's potential for wraparound to occur before the check happens, and it's a bit complicated to think about when different types are mixed, especially since `size_t` and `int64_t` are not necessarily ordered as one being bigger than the other, and they differ in signedness. At least we're not mixing 3 different data types together now though.

Since both `position` and `streamSize` are unsigned values (now of the same type), they might hold a large positive value just short of the maximum range, with `position` being less than `streamSize`. Using a small positive `offset`, greater than the number of bytes left available, the sum `position + offset` might overflow and wrap around to a small positive value (the result will be unsigned). That small positive value will almost certainly be less than the large positive value of `streamSize`. Thus a small relative seek of only a few bytes could reset a stream position from near the end back to near the beginning, assuming the stream size was quite large.

In practice, we won't be using stream sizes that are big enough for that to be an issue, and OS limitations may keep us from hitting that issue (unlikely you can allocate a buffer spanning half the address space or more), though we should at least comment this. In assembly it would be very easy to check overflow/carry flags, though in portable C++ it gets a little more complicated.

Reply:
I took a stab at catching the bounds issues. Will add to next commit.